source/common/common/key_value_store_base.cc

```diff
@@ -58,20 +70,31 @@ bool KeyValueStoreBase::parseContents(absl::string_view contents) {
   return true;
 }
 
-void KeyValueStoreBase::addOrUpdate(absl::string_view key_view, absl::string_view value_view) {
+void KeyValueStoreBase::addOrUpdate(absl::string_view key_view, absl::string_view value_view,
+                                    absl::optional<std::chrono::milliseconds> ttl) {
   ENVOY_BUG(!under_iterate_, "addOrUpdate under the stack of iterate");
   std::string key(key_view);
   std::string value(value_view);
+  // Remove ttl if <= 0
+  if (ttl && ttl <= std::chrono::milliseconds(0)) {
+    ASSERT(false);
+    ttl = absl::nullopt;
+  }
+
   // Attempt to insert the entry into the store. If it already exists, remove
   // the old entry and insert the new one so it will be in the proper place in
   // the linked list.
   if (!store_.emplace(key, value).second) {
     store_.erase(key);
     store_.emplace(key, value);
+    ttl_manager_.clear(key);
   }
   if (max_entries_ && store_.size() > max_entries_) {
     store_.pop_front();
   }
+  if (ttl) {
+    ttl_manager_.add(ttl.value(), key);
+  }
   if (!flush_timer_->enabled()) {
     flush();
   }
```

Comment:
I think we should also refresh the TTL (e.g. clear() followed by add()) when `get()` is called for a key.

Reply:
Oh!  I didn't know that's the behavior we were going for.

Reply:
Interesting idea, @abeyad. For alt-svc we definitely don't want to extend the ttl when we read from the store. Do you think we want this behavior for xDS?

Reply:
Talked to Ali offline, we think the current behavior will work for both the xds and alt-svc use-cases.

Reply:
Chatted w/ Chris offline just now as well, I agree we don't want to do this in the KV store.  I was thinking of the scenario where a resource doesn't have a TTL, so we won't have any eviction and if the resource stops getting referenced, we wouldn't get updates for it, so we'd just have an orphaned entry laying around.  But I think we can and should solve for this at a layer above the KV store APIs (in our custom xDS store that uses the KV store).

So in conclusion: ignore what I said :)

Reply:
Heh. Yeah, I think doing that at a layer above the KV store sounds like the right approach, too!